### PR TITLE
"*" Not working as a landing command, changed text accordingly

### PR DESCRIPTION
**Pull request by Maverick283**
Hope I got this all right... Issue #330 brought to attention that the "*" command is not a valid command for the ILS approach. I changed the tutorial to show both the "Land" command and its abbreviation, "l" to use for getting an aircraft locked on the ILS.

### CODE REVIEW

**Comment by erikquinn** [comment]
The asterisk **key** works, but doing shift+8 doesn't, because the asterisk key actually enters a more appealing star (Unicode 2B50) instead of a normal asterisk symbol.

To make shift+8 also work the same way, copy lines 234-238 from input.js and add another case "56" and an "if(e.shiftKey)" before the rest of the stuff. Or, just remove the case that creates the star, but that will also require changes to aircraft.js as well.

However, I'm pretty sure that "l28r" as amended will not work, as runLanding() probably splices by spaces. Unless that's not true, it should be "l 28r" instead of "l28r".

Sidenote: If we wanted to add 'i' and 'ils' to the runLanding synonyms, then it would be okay to have a tutorial direct them to use 'i 28r', but honestly we are clearing them for the ils approach, NOT clearing them to land, which is a distinction that has always irked me somewhat, so in my opinion, because of that issue, I'd really recommend against telling people to use 'land' at all, since we aren't clearing them to land, but rather clearing them for an approach. My thoughts anyway.

**Comment by Maverick283** [comment]
Yeah, you are definitely right, I added the space.

I think we should keep the tutorial to show the "land" command (or rather the "l" command) because of the problem you addressed. My asterisk (on a QWERTZ keyboard) is on the same button as the + Symbol. To use it, [Shift] + "+" needs to be pressed which results in zooming in and no input to the text field. Thus I tried copying a asterisk from somewhere else, and past it in the text box. I tried both, "\* 28L" and "*28L", but neither command was accepted.

Maybe using "keys" is the wrong approach on for this? Why not use symbols instead?  

As for the landing clearances... I brought this issue to attention in #262 but we came to the conclusion to leave it the way it is it would increase the workload for the controller a lot!

**Comment by erikquinn** [comment]
I reviewed #262, and while I agree with everything you were saying in it, I also have to also agree with what everyone else said there: requiring separate approach clearance and then landing clearance commands adds more workload than it's worth. I'm all for realism, but as far as using the sim for practice, making me enter both commands doesn't add value.

What I'm suggesting is that we continue with a single command, but we add 'ils' as a synonym for runLanding(), so from the user's perspective, they don't clear a/c to "land", but instead clear it for "ils". And the tutorial should guide people to then use 'i 28r' instead of 'l 28r' (as an alternative to the \* key). Since we are descending and vectoring our arrivals in an attempt to get them onto the localizer, it makes much more sense to present it this way, and ignore the need to explicitly give a landing clearance (as opposed to giving a landing clearance and ignoring the need to explicitly give an approach clearance). Based on what you were saying in #262, this should probably sound like an attractive resolution to you too I would think.

One more note: Now that I think of it, laptops almost never have a numpad, so its crazy to expect everybody to use the asterisk by the numpad. Thus, we definitely need to make shift+8 have the same behavior. I could get to that tomorrow at the earliest though as I am out of town for a funeral.

**Comment by Maverick283** [comment]
Okay, I'll look into that code later and see if I can add/change the "l" to an "i".

As for the Shift+8 thing: If it type Shift + 8 I get a "(", not an Asterisks, as the symbols are layouted different on a QWERTZ Keyboard, and I'm assuming French Keyboards are different then QWERTY and QWERTZ and so on. So what I'm suggesting is that either instead of the actual key to be detected (Shift + 8) the program detects the symbol that is put (Asterisk) when it is pressed. 
That way, when I do Shift + "+", I get the same behavior you get when pressing Shift + 8. 
Or we create a smart tutorial, that knows what symbol is typed when pressing Shift + 8 on the keyboard the user is using, and displays it, for example so the text would go: "In order for a plane to align with the ILS, you have to type the command "(28L" " for me and "In order for a plane to align with the ILS, you have to type the command "*28L" " for you...

**Comment by erikquinn** [comment]
Alright, I'm with ya now. in that case, we will have to remove the fancy Unicode star, which is totally fine because its only cosmetic. I honestly just didn't know there were such differences in keyboard layouts.

I made the changes so I know where all this stuff is already; I can just remove the fancy star and add 'i'/'ils'. It'll be a super quick fix, and then I'll post a link here to to see if its working for you and montypython

**Comment by erikquinn** [comment]
I made a new branch "fix-asterisk" and merged them into my gh-pages.

@Maverick283 @montypython112 When able, could you have a look at [http://erikquinn.github.io/atc/](http://erikquinn.github.io/atc/) and verify the "*28r" command works on your keyboard layout now?

**Comment by Maverick283** [comment]
@erikquinn, it works if I use my external keyboard, that has a numpad, yet when I use my laptop, it will still zoom in rather then writing a Asterisk. Because the + is on the same key. Pressing Shift + "+" results in zooming in, because it is the key that is detected.

**Comment by Maverick283** [comment]
Did you also implement the "i 28r" to work as an addition to the "l 28r"? If so, I'll rechange the tutorial and finally contribute some (very little) code to this whole thing rather then just my usual blablabla xD

**Comment by erikquinn** [comment]
@Maverick283 yeah, I added 'i 28r' up there. Now that I think about it though, it probably would be nice to get rid of the darn space and make it 'i28r'... regardless, there are bigger problems.
image
Assuming this is the keyboard layout you use, I see what you're talking about, and I _don't_ see an easy way to make it play nice with non-qwerty layouts. I thought maybe it could be fixed by stopping all use of [`e.which`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/which), and instead using [`e.key`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key), which references the _content_ of the key (ie '*') instead of the key code (ie '106'), because the latter is different on different layouts, BUT it turns out Chrome doesn't like `e.key` anyway.

Apparently, according to [the internet](http://stackoverflow.com/a/12825956/5774767), the inability to deal well with multiple keyboard layouts with key events is a pretty common problem that doesn't have an attractive solution... which I find really surprising.

Note: the reason it doesn't work on the German QWERTZ layout (pictured above), is that the `* / + / ~` key (left of enter) has a code 187 returned by e.which. While the US QWERTY layout, the key in that location shows `} / ]` and returns code 221, and the `+ / =` key (left of backspace) returns code 187. Thus, since the code is set to reset zoom on [code 187] and zoom in on [shift + code 187], it is actually on the german qwertz layout _resetting the zoom_ when the key labeled "+" is pressed, and zooming in when shift + "+" (aka the \* symbol) is pressed. Thus, the attempt to make the asterisk from the keypad (not numpad) work as well is impossible, since it is placed elsewhere in different layouts.

If only there was a way to reference the CHARACTER being pressed instead of dealing with key codes... this is getting beyond my expertise though, for sure. I wonder if any of the "big dogs" around here have advice on this?

**Comment by Maverick283** [comment]
I'm like a little chihuahua... But what if we don't use the Asterisk but another character that would work instead? For example the "@" symbol? (Think it is not used yet). As far as I know, it is [Shift] + [2] for QWERTY, and it is [Alt Gr] + [q] for QWERTZ. For QWERTY I think it is great since it is right next to the 1 & 3, so the numbers used most as the first digit when assigning a runway. Also it makes sense to have a aircraft land "at" a runway ;)

If the [Alt Gr] gives you trouble (I know this is unique to QWERTZ Layouts), then maybe the "@" isn't the best solution. Then "#" comes to mind, and so on...

**Comment by erikquinn** [comment]
@ would still have the same problem. Though we can both enter an "@" symbol, the code only cares about the "keyboard event", which will tell you the "code" of the key pressed.
- on qwerty: `[shift] + [2] --> code 50`
- on qwertz: `[alt gr] + [q] --> code 81`

Since we differentiate by the key code in input.js (line 171), the two @ entries are not the same. We could add both, but then my "q" key would enter an "@" symbol... it just gets messy. That's why we would want to be able to differentiate not by key codes but by the character that the key would type, but there doesn't seem to be a way to do so. We just haven't run into this problem yet because up to this point, we never used many designated keys.

Of course, this is all only a problem because I'm trying to eliminate the need to hit the spacebar (as it is way out of the way when using arrow-key controls, which I still strongly believe are the way to go). The idea is that you could type `fh250` and then strike the "*" key (wherever it may be), and the preceding space would be entered for you. Thus, to type `fh250 *28r`, you go [f] [h] [2] [5] [0] [*] [2] [8] [r] [enter], and the command goes in properly. If there's no good solution to this problem, maybe I would just have to deal with entering the space manually. It just seems insane that we can't get it to work this way.

And not at all! Keep bringing on the ideas; I want to see this problem solved just as much as the next guy :smile:

**Comment by Maverick283** [comment]
Okay, so we are basically looking for a key that has no usage at all at the moment...? What about "/"? It is on the numpad, a key that I don't believe is yet used on QWERTY, and would be Shift + 7 for QWERTZ. As I don't want to be egoistic, I started looking where the "/" is on other keyboards (AZERTY has it where  the QWERTY ">" is), and it gets worse the more keyboards there are to add...

Two solutions come to mind:
- Let the user select his layout and execute commands based on his selection
- Find a way to detect not the key but the symbol pressed

For the user selection, that could go hand in hand with a translation as suggested in #232.

And I will now create a new issue as this is getting out of hand here...

File: assets/scripts/tutorial.js [hunk 1]
```diff
@@ -292,7 +292,7 @@ function tutorial_init_pre() {
     text:     ["You may choose to enter one command at a time, but air traffic controllers usually do multiple. Particularly in approach clearances,",
                "they follow an acronym &ldquo;PTAC&rdquo; for the four elements of an approach clearance, the &lsquo;T&rsquo; and &lsquo;C&rsquo; of which",
                "stand for &lsquo;Turn&rsquo; and &lsquo;Clearance&rsquo;, both of which we entered separately in this tutorial. Though longer, it is both ",
-               "easier and more real-world accurate to enter them together, like this: &lsquo;fh250 *28r&rsquo;."
+               "easier and more real-world accurate to enter them together, like this: &lsquo;fh250 l28r&rsquo;."
```
Comment (tedrek) [inline assets/scripts/tutorial.js:295]: I believe this should include a space like `l 28r`

**Pull request merged**